### PR TITLE
Reduce boxing in callback structs

### LINKED ISSUE #513: Reduce boxing in callback structs
RocksDB expose several APIs to allow client code to inject custom logic, such as compaction filter and table properties collector. rust-rocksdb usually implement such APIs in the follow way. 

Assume rocksdb expose such a struct for callback
```
// C++
class Callback {
    virtual void SomeCallback() = 0;
}
```

rust-rocksdb usually provide the API in the following way:
```
// Rust
pub trait Callback {
    fn callback();
}

pub fn new_callback<C: Callback>(callback: C) -> *mut DBCallback {
    let boxed: Box<dyn Callback> = Box::new(callback) as _;
    unsafe {
        crocksdb_ffi::crocksdb_callback_create(
            Box::into_raw(Box::new(boxed)),
            callback_fn,
        )
    }
}

extern "C" fn callback_fn(ctx: *mut c_void) {
    let callback = &*(ctx as *mut Box<dyn Callback>);
    callback.callback();
}
```

Note here `ctx` is a `Box<Box<dyn Callback>>`. The inner box is to hold the trait object. The outer box is needed since C code cannot hold the raw pointer of `Box<dyn Callback>` natively since the pointer is of 128-bits. However, the extra boxing can be reduced by using generic.

```
pub fn new_callback<C: Callback>(callback: C) -> *mut DBCallback {
    let boxed: Box<C> = Box::new(callback);
    unsafe {
        crocksdb_ffi::crocksdb_callback_create(
            Box::into_raw(boxed),
            callback_fn::<C>,
        )
    }
}

extern "C" fn callback_fn<C: Callback>(ctx: *mut c_void) {
    let callback = &*(ctx as *mut C);
    callback.callback();
}
```

The following APIs need to be updated:
* CompactionFilter
* TablePropertiesCollector
* SliceTransform
* EventListener
* Logger
* TableFilter

### PR DESCRIPTION
**Pull request by yuqi1129**
Signed-off-by: yuqi1129 <yuqi4733@gmail.com>

This PR tries to resolve issue https://github.com/tikv/rust-rocksdb/issues/513

Close #513.

### CODE REVIEW

File: src/event_listener.rs [hunk 1]
```diff
@@ -355,16 +355,16 @@ pub fn new_event_listener<L: EventListener>(l: L) -> *mut DBEventListener {
...
-            on_flush_begin,
-            on_flush_completed,
-            on_compaction_begin,
-            on_compaction_completed,
-            on_subcompaction_begin,
-            on_subcompaction_completed,
-            on_external_file_ingested,
-            on_background_error,
-            on_stall_conditions_changed,
+            destructor::<L>,
```
Comment (Connor1996) [inline src/event_listener.rs:358]: To keep them consistent, let's use `E` to denote the type

File: src/logger.rs [hunk 2]
```diff
@@ -10,15 +10,15 @@ pub trait Logger: Send + Sync {
     fn logv(&self, log_level: InfoLogLevel, log: &str);
 }
 
-extern "C" fn destructor(ctx: *mut c_void) {
+extern "C" fn destructor<C: Logger>(ctx: *mut c_void) {
```
Comment (Connor1996) [inline src/logger.rs:13]:
```suggestion
extern "C" fn destructor<L: Logger>(ctx: *mut c_void) {
```

File: src/table_properties_collector_factory.rs [hunk 3]
```diff
@@ -14,65 +14,81 @@
...
-struct TablePropertiesCollectorFactoryHandle {
+struct TablePropertiesCollectorFactoryHandle<C, T>
+where
+    C: TablePropertiesCollector,
+    T: TablePropertiesCollectorFactory<C>,
+{
     name: CString,
-    rep: Box<dyn TablePropertiesCollectorFactory>,
+    rep: T,
+    _marker: PhantomData<C>,
```
Comment (tabokie) [inline src/table_properties_collector_factory.rs:34]: Maybe `_phantom`?

**Review by tabokie — APPROVED** [review]
LGTM

**Review by tabokie — COMMENTED** [review]
What's changed in the rocksdb submodule? Maybe that caused CI failure.

**Comment by yuqi1129** [comment]
@tabokie ，Seems that it's the failure in master branch that make this PR fail..

**Comment by tabokie** [comment]
> @tabokie ，Seems that it's the failure in master branch that make this PR fail..

If you are referring to the branch status, it's a no-op and we usually ignore it.

I think you are missing this PR: https://github.com/tikv/rust-rocksdb/pull/661, the rocksdb submodule is not in sync with master branch. You can probably update it with `git submodule update --remote`.

**Comment by yuqi1129** [comment]
> > @tabokie ，Seems that it's the failure in master branch that make this PR fail..
> 
> If you are referring to the branch status, it's a no-op and we usually ignore it.
> 
> I think you are missing this PR: #661, the rocksdb submodule is not in sync with master branch. You can probably update it with `git submodule update --remote`.

You mean we can just fix the UT by execute `git submodule update --remote`, i have done this locally, but problem still exists, correct me if i'am wrong.

**Comment by tabokie** [comment]
@yuqi1129 I'm not sure what went wrong at your end, but you did changed the rocksdb folder in this [commit](https://github.com/tikv/rust-rocksdb/pull/663/commits/6bdaa5c6c287cf787c8c8add06b385a0bcb3d93c), so this PR effectively reverts https://github.com/tikv/rocksdb/pull/257, and break the test newly introduced in https://github.com/tikv/rust-rocksdb/pull/662.

**Comment by yiwu-arbug** [comment]
@yuqi1129 does `cargo test` pass on your side?

**Pull request merged**